### PR TITLE
improve website design for github pages

### PR DESCRIPTION
**Pull request by initbar**
### Before submitting a *pull request* make sure you have:
- [X] At least skimmed through [adding new extractor tutorial](https://github.com/rg3/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/rg3/youtube-dl#youtube-dl-coding-conventions) sections.
- [X] [Searched](https://github.com/rg3/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests.

### What is the purpose of your *pull request*?
- [x] Bug fix
- [ ] New extractor
- [ ] New feature
- [X] gh-pages website design improvement (issue #1025, #2440)

### Description of your *pull request* and other information

This pull request does the followings:

- organize assets/codes into clean layout:
~~~
root --- static/
     \-- update/
     \-- donations.html
     \-- index.html
     \-- installation.html
     \-- installation.in
     \-- supported.html
     \-- supported.in
~~~
- utilize bootstrap and awesome icons.
- migrate from pure-css rules to `sass` syntax (generating css rules can be done using `sass -C --style compressed <style.sass>`).
- improve main page.
- improve supported websites page.
- merge multiple files into one design.
- added a temporary favicon.

### Todo

- Improved mobile support.
- Fix unknown spacing for `<pre><code>{{ content }}</code></pre>`.
- Organize sass file (and remove possible duplicate/unnecessary rules).

### Note

I realize there are still components which requires further improvements, but staging this commit first was more important than proceeding. If there are any problems/questions with this pull request, please let me know and I'll resolve them asap.

### CODE REVIEW

**Comment by yan12125** [comment]
Some quick ideas after comparing the old and the new one:
* The "CC-BY-SA Copyright © 2006-2016 Ricardo Garcia Gonzalez" line at the bottom should be preserved.
* It's better to have some words mentioning GPG keys and SHA256 checksums in the download section. It's important that users have downloaded the correct file.

**Comment by yan12125** [comment]
And, the recommended installation method on Unix-like systems is not ```git clone``` but wget/curl/pip/Homebrew.

**Comment by phihag** [comment]
URLs must not change. I strongly recommend moving the change from `supportedsited` to `supported` into a separate pull request.

**Comment by phihag** [comment]
The proposed changes come with major privacy implications; referencing files from third-party hosts. I strongly oppose this, as it allows tracking users. Also, attackers who compromise these hosts could inject attacks or change content on our website.

File: index.html [hunk 1]
```diff
@@ -1,37 +1,121 @@
...
+<div class="table">
+  <div class="t-body">
+    <div class="t-row">
+      <div class="t-cell">
+        <i class="fa fa-code" aria-hidden="true"></i>
+        <h1>Python 2/3+</h1>
+        <p>Purely written in Python from scratch.</p>
+      </div>
+      <div class="t-cell">
+        <i class="fa fa-terminal" aria-hidden="true"></i>
```
Comment (phihag) [inline index.html]: This looks like dead code to me, and using a deprecated element as well.

**Comment by phihag** [comment]
The CSS class names throughout the proposed versions are often meaningless or effectively [declaring style, not semantics](https://www.w3.org/QA/Tips/goodclassnames).

File: installation.html [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+<header class="container-fluid">
+  <div class="container vertical-center">
+    <h1>Installation Manual</h1>
+  </div>
+</header>
+
+<div id="manual" class="container">
+
+  <p>Note that if you directly click on these hyperlinks, your web browser will most likely display the program contents. It's usually better to right-click on it and choose the appropriate option, normally called <em>Save Target As</em> or <em>Save Link As</em>, depending on the web browser you are using.</p>
```
Comment (phihag) [inline installation.html]: If any web browser indeed does display program contents, we may want to set better attributes and/or HTTP headers.
Can you name one or more browsers which do this? It sounds as though you assume it is not a bug in a single web browser, say IE6.

**Comment by phihag** [comment]
The sass generation still needs to be automated - GitHub pages renders what's in git. A `.gitignore` entry would also be pertinent.

**Comment by phihag** [comment]
Why not just put the favicon it in `/favicon.ico`? That way, we don't have to reference it.

**Comment by phihag** [comment]
There is no source file for the favicon. Would you be willing to contribute the original SVG?

**Comment by phihag** [comment]
Thank you very much for your contribution! However, due to the numerous issues mentioned above, I am afraid that I have to reject this pull request.

I don't think there's anything wrong with the current page, although we may want to remove the superfluous background. So I'd rather go an evolutionary way and recommend to file multiple smaller PRs, each modifying one aspect.

I think some of your additions, namely favicon, mobile support and improved wording, are indeed great ideas, and would be happy to see them in further PRs.

**Comment by phihag** [comment]
I also have to say that I do like the proposed style of the /index page and would be very happy to see something to this effect in the future, maybe a little bit more compressed so all the important information fits onto the screen without having to scroll.

**Comment by initbar** [comment]
There are numerous points made in here, so I'll try my best to address them all :sweat_smile:.

> Some quick ideas after comparing the old and the new one:
>
> * The "CC-BY-SA Copyright © 2006-2016 Ricardo Garcia Gonzalez" line at the bottom should be preserved.
> * It's better to have some words mentioning GPG keys and SHA256 checksums in the download section. It's important that users have downloaded the correct file.

@yan12125 I'll re-apply `CC-BY-SA Copyright © 2006-2016 Ricardo Garcia Gonzalez` at the footer of *index.html* and double check so that it is ubiquitous to all pages. I'll also look into adding file verification/checksums in the downloads section.

> And, the recommended installation method on Unix-like systems is not git clone but wget/curl/pip/Homebrew.

@yan12125 It'll be overwhelming to see wget/curl/pip/Homebrew guides blobbed together. Furthermore these methods are specific to Linux distros and OSX (e.g. Homebrew). Instead, how about something like this?

```html
<p>
  Please see installation guide for <a href="./supported.html#wget">wget</a>, <a href="./supported.html#curl">curl</a>, <a href="./supported.html#pip">pip</a>, or <a href="./supported.html#homebrew">Homebrew</a>.
</p>
```

Or simply redirect each links to to this project's [installation guide](https://github.com/rg3/youtube-dl#installation).

> URLs must not change. I strongly recommend moving the change from supportedsited to supported into a separate pull request.

@phihag Understood, they will be on a separate PR request.

> The proposed changes come with major privacy implications; referencing files from third-party hosts. I strongly oppose this, as it allows tracking users. Also, attackers who compromise these hosts could inject attacks or change content on our website.

@phihag I can come up with two solutions right now (based on limitations of Github Pages):

1. Cache `bootstrap.min.css` and `font-awesome.min.css` locally into `static/` folder (although fonts must be explicitly referenced to Google - unless you're fine with adding fonts' weight to the `gh-pages` branch).

2. Although this addresses the security aspect, add [SRI](https://developer.mozilla.org/en-US/docs/Web/Security/Subresource_Integrity) to third party stylesheets. Or, tap into the CSP's content integrity rules (e.g. hashes).

> This looks like dead code to me, and using a deprecated element as well.

@phihag I do see some parts can be further reduced - but the transition from `<table>` to `<div>` (for the sake of responsive tables) is [vital](http://programmers.stackexchange.com/a/277985) and requires heavy nesting of `<div>` elements. Nonetheless, what was the deprecated element that you've noted? I've double checked and all elements are handled in HTML5?

> The CSS class names throughout the proposed versions are often meaningless or effectively declaring style, not semantics.

@phihag Agreed. This will be fixed in the next PR.

> If any web browser indeed does display program contents, we may want to set better attributes and/or HTTP headers.
Can you name one or more browsers which do this? It sounds as though you assume it is not a bug in a single web browser, say IE6.

@phihag There may have been a confusion here.. The text regarding displaying program contents was actually from [here](https://rg3.github.io/youtube-dl/download.html) (not an assumption that I have made). When I was redesigning the Downloads page, I also noticed modern browsers no longer does this. Would you prefer [this line](https://github.com/rg3/youtube-dl/pull/10265/files#diff-518ca5fb7da42aa3de8d6009c13c1ac7R25) be removed?

> The sass generation still needs to be automated - GitHub pages renders what's in git. A .gitignore entry would also be pertinent.

@phihag Got it - I can write a small bash script which regenerates css rules from `style.sass` per second and overwrite the existing `style.css`. I'll also add an entry to the `.gitignore` file :smile:.

> Why not just put the favicon it in /favicon.ico? That way, we don't have to reference it.
>
> There is no source file for the favicon. Would you be willing to contribute the original SVG?

@phihag The favicon was actually purposed for a temporary placeholder drafted using [favicon.cc](http://www.favicon.cc/). It's definitely not worthy for the youtube-dl project.

> I also have to say that I do like the proposed style of the /index page and would be very happy to see something to this effect in the future, maybe a little bit more compressed so all the important information fits onto the screen without having to scroll.

@phihag To clarify, would you prefer the website to have a minimalist homepage and link to its' multiple components (like how it's layout currently)?

**Comment by yan12125** [comment]
> It'll be overwhelming to see wget/curl/pip/Homebrew guides blobbed together.

Maybe it's time to determine a recommended tool and list others as alternatives.

**Comment by phihag** [comment]
>     It'll be overwhelming to see wget/curl/pip/Homebrew guides blobbed
>     together.
> Maybe it's time to determine a recommended tool and list others as
> alternatives.

Homebrew is basically macOS-only, so that can't be it (plus, they do
downstream packaging). It is familiar to macOS users though - that's why
it's mentioned.

pip is not installed on any systems out of the box. It's the only CLI
option that regularily works on Windows though (although that may
change). It's also familiar to Python developers.

wget is the only one in the list available on all debian systems, but
it's not installed in macOS and Arch Linux.

On Arch Linux and macOS, curl is available out of the box.

For these reasons, I don't think there's any one we can remove.

**Comment by phihag** [comment]
> @phihag I can come up with two solutions right now (based on limitations of Github Pages):
> Cache bootstrap.min.css (...)

Do we really need that huge CSS file?

> @phihag I do see some parts can be further reduced - but the transition from <table> to <div> (for the sake of responsive tables) is vital and requires heavy nesting of <div> elements.

`<div>`s are not vital, [your link](http://programmers.stackexchange.com/questions/277778/why-are-people-making-tables-with-divs/277985#277985) says so at the end and in the comments as well. (Also look at the top-voted answer, which is imho waaay better).

> Nonetheless, what was the deprecated element that you've noted? I've double checked and all elements are handled in HTML5?
Oh, my mistake, no deprecated stuff there, except repeating visual styles, disguised as CSS classes.
 
> @phihag There may have been a confusion here.. The text regarding displaying program contents was actually from here (not an assumption that I have made). When I was redesigning the Downloads page, I also noticed modern browsers no longer does this. Would you prefer this line be removed?
I have jsut removed it.

> The sass generation still needs to be automated - GitHub pages renders what's in git. A .gitignore entry would also be pertinent.
> @phihag Got it - I can write a small bash script which regenerates css rules from style.sass per second and overwrite the existing style.css. I'll also add an entry to the .gitignore file .

Or simply compile the sass to CSS and only check that in? Maybe I'm a dinosaur, but for the simplicity.

> I also have to say that I do like the proposed style of the /index page and would be very happy to see something to this effect in the future, maybe a little bit more compressed so all the important information fits onto the screen without having to scroll.
> @phihag To clarify, would you prefer the website to have a minimalist homepage and link to its' multiple components (like how it's layout currently)?

No, I do think your suggested style is superior. We can seriously think of putting all the information on one page to speed up navigation (maybe except the supportedsites). I just think that a 90% red page (that's how it looks for me on monitor that's probably not the smallest of our users at 2560x1440) with little information may look cool, but is not as functional or usable as a short introduction.
No, I do think your suggested style is superior. We can seriously think of putting all the information on one page to speed up navigation. I just think that a 90% red page (that's how it looks for me on monitor that's probably not the smallest of our users at 2560x1440) with little information may look cool, but is not as functional or usable as a short introduction.

**Pull request closed without merging**